bluesky/callbacks/core.py

```diff
@@ -540,3 +541,139 @@ def event(self, doc):
+        #E 1449179338.3418093
+        #D 2015-12-03 16:48:58.341809
+        #C xf11id  User = xf11id
+        #O [list of all motors, all on one line]
+        session_manager = get_session_manager()
+        pos = session_manager.get_positioners()
+        spec_header = [
+            '#F %s' % self.specpath,
+            '#E %s' % int(doc['time']),
+            # time might need to be formatted specifically
+            '#D %s' % datetime.fromtimestamp(doc['time']),
+            '#C %s  User = %s' % (doc['owner'], doc['owner']),
+            'O0 {}'.format(' '.join(sorted(list(pos.keys()))))
+        ]
+        with open(self.specpath, 'w') as f:
+            f.write('\n'.join(spec_header))
+
+        return spec_header
+
+    def start(self, doc):
+        from ophyd.session import get_session_manager
+        from IPython import get_ipython
+
+        if not os.path.exists(self.specpath):
+            spec_header = self._write_spec_header(doc)
+        # grab the literal last command that the user typed at the command line
+        # in IPython and format it to remove the parentheses and commas
+        # e.g. dscan(th, 0, 1, 10, .1) -> dscan th 0 1 10 .1
+        last_command = list(
+            get_ipython().history_manager.get_range())[-1][2]
+        last_command = last_command.replace('(', ' ')
+        last_command = last_command.replace(')', ' ')
+        last_command = last_command.replace(',', ' ')
+        dets = eval(doc['detectors'])
+        # grab the acquisition time from one of the detectors.  This part does
+        # not need to be changed if the detectors do not have the same
+        # acquisition time. acquisition_time is just one of those things that
+        # SPEC needs
+        self.acquisition_time = dets[0].acquire_time
+        # write a blank line between scans
+        with open(self.specpath, 'a') as f:
+            f.write('\n\n')
+        # write the new scan entry
+        with open(self.specpath, 'a') as f:
+            f.write('#S %s %s\n' % (doc['scan_id'], last_command))
+            f.write('#D %s\n' % datetime.fromtimestamp(doc['time']))
+            f.write('#T %s (Seconds)\n' % self.acquisition_time)
+        # write the motor positions
+        pos = get_session_manager().get_positioners()
+        positions = [str(v.position) for k, v in sorted(pos.items())]
+        with open(self.specpath, 'a') as f:
+            f.write('#P0 {0}\n'.format(' '.join(positions)))
+        # writing the list of motor names and their current values in
+        # a more human readable way. Apparently "#M" is not a used key in
+        # spec. Fingers crossed....
+        with open(self.specpath, 'a') as f:
+            for idx, (name, positioner) in enumerate(sorted(pos.items())):
+                f.write('#M%s %s %s\n' % (idx, name, str(positioner.position)))
+        print("RunStart document received in LiveSpecFile")
+        #raise
+        self.motorname = eval(doc['motor']).name
```

Comment:
I'm not saying this is pretty, sensibly named, or how you should do it, but I mocked up "`NamedObjects`" to extract object names from bluesky `scan_args` in hxntools like so:
```python
def _eval(scan_args):
    '''Evaluate scan arguments, replacing OphydObjects with NamedObjects'''

    class NamedObject:
        def __init__(self, name):
            self.name = name

    def no_op():
        def no_op_inner(*args, name=None, **kwargs):
            if name is not None:
                return NamedObject(name)

        return no_op_inner

    return eval(scan_args, collections.defaultdict(no_op))
```